### PR TITLE
Core

### PR DESCRIPTION
**Pull request by ZeeshanGhalib**

### CODE REVIEW

File: core/src/main/java/uk/gov/justice/services/core/annotation/Adapter.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Identifies adapters that receive messages from other service components.
+ * <p>
+ * Usage: @Adapter({@link Component#COMMAND_CONTROLLER})
+ */
+
+@Retention(RetentionPolicy.RUNTIME)
+@ApplicationScoped
+public @interface Adapter {
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/annotation/Adapter.java]: Does this need a @Target annotation to limit it to classes?

File: core/src/main/java/uk/gov/justice/services/core/annotation/Component.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    COMMAND_API, COMMAND_CONTROLLER, COMMAND_HANDLER;
+
+    /**
+     * Retrieves the Component of the provided {@link ServiceComponent}
+     *
+     * @param clazz The service component to be analysed.
+     * @return the component from the provided {@link ServiceComponent}.
+     */
+    public static Component getComponentFromServiceComponent(Class<ServiceComponent> clazz) {
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/annotation/Component.java]: Clazz argument should be final

File: core/src/main/java/uk/gov/justice/services/core/annotation/ServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Identifies service components.
+ * <p>
+ * Usage: @ServiceComponent({@link Component#COMMAND_API})
+ */
+
+@Retention(RetentionPolicy.RUNTIME)
+@ApplicationScoped
+public @interface ServiceComponent {
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/annotation/ServiceComponent.java]: Again, I think this should have an @Target

File: core/src/main/java/uk/gov/justice/services/core/dispatcher/AsynchronousDispatcher.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import uk.gov.justice.services.core.handler.registry.HandlerRegistry;
+import uk.gov.justice.services.messaging.Envelope;
+
+/**
+ * Dispatches messages asynchronously to their corresponding handlers, which could be a command handler,
+ * command controller, event processor, etc.
+ */
+public class AsynchronousDispatcher implements Dispatcher {
+
+    HandlerRegistry handlerRegistry;
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/dispatcher/AsynchronousDispatcher.java]: Should this be private? Or protected?

File: core/src/main/java/uk/gov/justice/services/core/dispatcher/DispatcherProducer.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.inject.Inject;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@ApplicationScoped
+public class DispatcherProducer {
+
+    Map<Component, HandlerRegistry> handlerRegistryMap;
+    Map<Component, Dispatcher> dispatcherMap;
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/dispatcher/DispatcherProducer.java]: These should probably be private

File: core/src/main/java/uk/gov/justice/services/core/extension/AnnotationScanner.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * Returns an event based on the handler type of the bean.
+     *
+     * @param bean a bean that could be a handler.
+     * @return an optional event.
+     */
+    private Optional<ServiceComponentFoundEvent> getEvent(final Bean bean) {
+        if (bean.getBeanClass().isAnnotationPresent(ServiceComponent.class)) {
+            return Optional.of(new ServiceComponentFoundEvent(Component.getComponentFromServiceComponent(bean.getBeanClass()), bean));
+        } else return Optional.empty();
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/extension/AnnotationScanner.java]: Sonarqube will complain about not using curly brackets here

File: core/src/main/java/uk/gov/justice/services/core/extension/ServiceComponentFoundEvent.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package uk.gov.justice.services.core.extension;
+
+import uk.gov.justice.services.core.annotation.Component;
+
+import javax.enterprise.inject.spi.Bean;
+
+public class ServiceComponentFoundEvent {
+
+    private Component component;
+    private Bean handlerBean;
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/extension/ServiceComponentFoundEvent.java]: these can be final

File: core/src/main/java/uk/gov/justice/services/core/util/JsonObjectConverter.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        return builder.build();
+    }
+
+    public JsonObject extractPayloadFromEnvelope(final JsonObject envelope) {
+        JsonObjectBuilder builder = Json.createObjectBuilder();
+        envelope.keySet().stream().filter(key -> !METADATA.equals(key)).forEach(key -> builder.add(key, envelope.get(key)));
+        return builder.build();
+    }
+
+}
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/util/JsonObjectConverter.java]: As discussed, we will probably want to convert this into something more like a "message converter". But we can do that as part of the JMS listener work.

File: core/src/test/java/uk/gov/justice/services/core/handler/registry/HandlerRegistryTest.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    private final static String COMMAND_NAME = "test.commands.mock-command";
+
+    @Mock
+    private TestCommandHandler commandHandler;
+
+    @Mock
+    private TestCommandHandlerDuplicate commandHandlerDuplicate;
+
+    @Mock
+    private TestCommandHandler commandHandlerPOJO;
```
Comment (purple52) [inline core/src/test/java/uk/gov/justice/services/core/handler/registry/HandlerRegistryTest.java]: Is this used anywhere?

File: core/src/main/java/uk/gov/justice/services/core/util/JsonObjectConverter.java (hunk 8 above)
Comment (ZeeshanGhalib) [inline core/src/main/java/uk/gov/justice/services/core/util/JsonObjectConverter.java]: Yes, as we agreed, we will do this as part of the JMS listener.

**Pull request merged**